### PR TITLE
Clean up the pixel vertex reconstruction code

### PR DESCRIPTION
**Pull request by ericcano**
#### PR description:

This PR will implement remarks for https://github.com/cms-sw/cmssw/pull/31723.

#### PR validation:

The baseline and new code will be run through TTbar validations:

   -  [501: CPU - Quad+Pentaplets](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.501-step3.py)
   -  [502: GPU - Quad+Pentaplets](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.502-step3.py)

  - [505: CPU - Trips+Quads+Penta](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.505-step3.py)
   - [506: GPU - Trips+Quads+Penta](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.506-step3.py)

### CODE REVIEW

File: RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc [hunk 1]
```diff
@@ -127,6 +129,7 @@ void PixelVertexProducerFromSoA::produce(edm::StreamID streamID, edm::Event &iEv
         edm::LogWarning("PixelVertexProducer") << "oops track " << it << " does not exists on CPU " << k;
         continue;
       }
+      auto trackCollection = iEvent.getHandle(tokenTracks_);
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:132]: This is cheap, but not free.
Could you move it back outside of the loop ?

File: RecoPixelVertexing/PixelVertexFinding/plugins/gpuVertexFinderImpl.h [hunk 2]
```diff
@@ -112,37 +118,42 @@ namespace gpuVertexFinder {
     init(soa, ws_d.get());
     loadTracks(tksoa, soa, ws_d.get(), ptMin);
 #endif
-
+    
 #ifdef __CUDACC__
+    // Running too many thread lead to problems when printf is enabled.
+    constexpr int maxThreadsForPrint = 1024 - 256;
+    constexpr int splitBlocks = 1024;
+    constexpr int splitThreadsPerBlock = 128;
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/plugins/gpuVertexFinderImpl.h:126]: could you use the names
```suggestion
    constexpr int numBlocks = 1024;
    constexpr int threadsPerBlock = 128;
```
?

File: RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc (hunk 1 above)
Comment (ericcano) [reply RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc]: Sure, fixed in latest commits.

File: RecoPixelVertexing/PixelVertexFinding/plugins/gpuVertexFinderImpl.h (hunk 2 above)
Comment (ericcano) [reply RecoPixelVertexing/PixelVertexFinding/plugins/gpuVertexFinderImpl.h]: Sure, fixed in latest commits.

File: RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc [hunk 3]
```diff
@@ -89,9 +87,12 @@ void PixelVertexProducerFromSoA::produce(edm::StreamID streamID, edm::Event &iEv
 
   int nv = soa.nvFinal;
 
-  // std::cout << "converting " << nv << " vertices " << " from " << indToEdm.size() << " tracks" << std::endl;
+#ifdef PIXVERTEX_DEBUG_PRODUCE
+  std::cout << "converting " << nv << " vertices " << " from " << indToEdm.size() << " tracks" << std::endl;
+#endif // PIXVERTEX_DEBUG_PRODUCE 
 
   std::set<uint16_t> uind;  // fort verifing index consistency
+  auto trackCollection = iEvent.getHandle(tokenTracks_);
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:95]: sorry, rereading the changes, isn't this the same as `tracks` at line 67 ?

Comment (ericcano) [reply RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc]: Indeed. 

We absolutely need a handle for making a TrackRef, as edm::Ref does not take references to product as a constructor parameter. It does does take pointers to product (and this compiles):
```c++
      //(old) auto tk = reco::TrackRef(trackCollectionHandle, k);
      auto tk = reco::TrackRef(&tracks, k);
```

but the constructor comments advise against using this to persist products which are in events (`DataFormats/Common/interface/Ref.h`):
```c++
    /// Constructors for ref to object that is not in an event.
    //  An exception will be thrown if an attempt is made to persistify
    //  any object containing this Ref.  Also, in the future work will
    //  be done to throw an exception if an attempt is made to put any object
    //  containing this Ref into an event(or run or lumi).
    Ref(product_type const* product, key_type itemKey, bool setNow = true);
```

So we can replace `track` with a handle and dereference it where needed (only one place, actually).

Fixed in latest commit, with an extra (unrelated) handle creation simplification.

Comment (fwyzard) [reply RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc]: > It does take pointers to product (and this compiles)

Sure, but as you have seen in the comment that will give a transient-only `edm::Ref`, which does not know the original collection - so it's not what we want.

File: RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc [hunk 4]
```diff
@@ -62,15 +64,9 @@ void PixelVertexProducerFromSoA::fillDescriptions(edm::ConfigurationDescriptions
...
-  edm::Handle<reco::TrackCollection> trackCollection;
-  iEvent.getByToken(tokenTracks_, trackCollection);
-  auto const &tracks = *(trackCollection.product());
-  edm::Handle<IndToEdm> indToEdmH;
-  iEvent.getByToken(tokenIndToEdm_, indToEdmH);
-  auto const &indToEdm = *indToEdmH;
-
-  edm::Handle<reco::BeamSpot> bsHandle;
-  iEvent.getByToken(tokenBeamSpot_, bsHandle);
+  auto tracksHandle = iEvent.getHandle(tokenTracks_);
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:67]:
```suggestion
  auto tracksHandle = iEvent.getHandle(tokenTracks_);
  const auto tracksSize = tracksHandle->size();
```

File: RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc [hunk 5]
```diff
@@ -123,11 +124,11 @@ void PixelVertexProducerFromSoA::produce(edm::StreamID streamID, edm::Event &iEv
     for (auto it : itrk) {
       assert(it < int(indToEdm.size()));
       auto k = indToEdm[it];
-      if (k > tracks.size()) {
+      if (k > tracksHandle->size()) {
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:127]:
```suggestion
      if (k > tracksSize) {
```

**Review by fwyzard — COMMENTED** [review]
Last minor suggestion.

File: RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc (hunk 5 above)
Comment (ericcano) [reply RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc]: Indeed.

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-37-02:/data/user/fwyzard/patatrack/validation/run_609.720XqNxPB7 ...

**Pull request merged**